### PR TITLE
Add `share` and `wrap` interfaces

### LINKED ISSUE #10: Exposing `dlpack` interface for Julia `Array`/`CuArray`
I think the current code only handles the `from_dlpack` via `unsafe_wrap` but not the `to_dlpack` part.

### PR DESCRIPTION
**Pull request by pabloferz**
Fixes #10

TODO

 - [x] Add CUDA support
 - [x] Add PythonCall support
 - [x] Add tests
 - [x] Remove `DLArray`

### CODE REVIEW

File: src/pycall.jl [hunk 1]
```diff
@@ -43,3 +51,21 @@ end
 function DLArray{T, N}(::Type{A}, ::Type{M}, o::PyObject, to_dlpack) where {T, N, A, M}
     return DLArray{T, N}(A, M, DLManagedTensor(to_dlpack(o)), o)
 end
+
+function to_dlpack(A::StridedArray, foreign_from_dlpack::Union{PyObject, Function})
```
[the remark this answers is not in the document]
Comment (pabloferz) [reply src/pycall.jl:55]: I must say I don't like the names `from_dlpack` and `to_dlpack` that much, but I was just following the convention.

If we were to change this I'd prefer something more explicit like `share` for exporting and `wrap` for importing.

[the remark this answers is not in the document]
Comment (rejuvyesh) [reply src/pycall.jl:55]: Also I think we need not be limited to `StridedArrays` and can support a wider type as long as `strides` is `applicable` as done by https://github.com/mkitti/NumPyArrays.jl/blob/main/src/NumPyArrays.jl#L87-L114

[the remark this answers is not in the document]
Comment (pabloferz) [reply src/pycall.jl:55]: The issue with that is not all python libraries that support importing via dlpack support general strides. I know least `jax` only supports compact strides.

[the remark this answers is not in the document]
Comment (pabloferz) [reply src/pycall.jl:55]: Still, I will check how much support can be expanded to a wider range of `AbstractArrays`.

[the remark this answers is not in the document]
Comment (rejuvyesh) [reply src/pycall.jl:55]: At worst I can just copy and make it strided, kind of what happens with [`torch.tensor(x).contiguous()`](https://pytorch.org/docs/stable/generated/torch.Tensor.contiguous.html)

**Comment by rejuvyesh** [comment]
```
signal (11): Segmentation fault
in expression starting at REPL[1]:1
unknown function (ip: 0x7f31c58049d0)
Allocations: 107518953 (Pool: 107475126; Big: 43827); GC: 77
```
I'm seeing intermittent segfaults like above with this branch, just so you are aware.

**Comment by pabloferz** [comment]
Do you have a reproducible example?

**Comment by rejuvyesh** [comment]
It's just intermittent unfortunately. With https://gist.github.com/rejuvyesh/0c0995ac81d8c75efada7797a292f611:

```julia
julia> include("test/stresstest_dlpack.jl")
[ Info: Precompiling Zygote [e88e6eb3-aa80-5325-afca-941959d7151f]
┌ Warning: `vendor()` is deprecated, use `BLAS.get_config()` and inspect the output instead
│   caller = npyinitialize() at numpy.jl:67
└ @ PyCall ~/.julia/packages/PyCall/L0fLP/src/numpy.jl:67
Test Passed
  Expression: size(grad.params[2]) == size(modelwrap.params[2])
   Evaluated: (2,) == (2,)

julia> include("test/stresstest_dlpack.jl")
Test Passed
  Expression: size(grad.params[2]) == size(modelwrap.params[2])
   Evaluated: (2,) == (2,)

julia> include("test/stresstest_dlpack.jl")
Test Passed
  Expression: size(grad.params[2]) == size(modelwrap.params[2])
   Evaluated: (2,) == (2,)

julia> include("test/stresstest_dlpack.jl")

signal (11): Segmentation fault
in expression starting at /home/jagupt/.julia/dev/PyCallChainRules/test/stresstest_dlpack.jl:102
unknown function (ip: 0x7f7076f69e30)
Allocations: 62477905 (Pool: 62457694; Big: 20211); GC: 69
```

**Comment by rejuvyesh** [comment]
Also I haven't been able to reproduce the segfault with `jax`'s `dlpack`.

**Comment by pabloferz** [comment]
Can you try again @rejuvyesh?

**Comment by rejuvyesh** [comment]
Yep, this fixes the issues! Amazing sleuthing!

**Comment by pabloferz** [comment]
To support both PyCall and PythonCall, I need to change the signature of `share` to something like `share(::StridedArray, ::Type{PyObject}, from_dlpack)` and `share(::StridedArray, ::Type{Py}, from_dlpack)` (`share(::StridedArray, from_dlpack::PyObject)` and `share(::StridedArray, from_dlpack::Py)` would still work though).

**Comment by rejuvyesh** [comment]
It seems we will need a separate wrapper for CUDA arrays than `DLArray` to avoid the `AbstractArray` fallback: https://github.com/JuliaLang/julia/pull/31563

**Comment by pabloferz** [comment]
Ok. I decided to write a `wrap` method to replace `DLArray` this method returns `Array`s and `CuArray`s directly and stores the python objects in a global pool instead. This should avoid any issues with `CUDA.allowscalar(false)`.

The only thing is that I dropped the `PermuteDimsArrays` wrappers for now (but I could added them if really necessary).

So now the interface looks like:

 - `DLPack.wrap` (similar to `from_dlpack` in python libraries)
 - `DLPack.share` (similar to `to_dlpack` in python libraries)

**Comment by pabloferz** [comment]
Once all python libraries support exporting and importing via `__dlpack__` attributes we can write `to_dlpack` and `from_dlpack` methods that match the semantics from https://data-apis.org/array-api/latest/design_topics/data_interchange.html. But I think `DLPack.wrap` and `DLPac.share` should work fine for now.

File: src/pycall.jl [hunk 2]
```diff
@@ -43,3 +48,35 @@ end
...
+    return wrap(DLManagedTensor(to_dlpack(o)), o)
+end
+#
+function wrap(::Type{A}, ::Type{M}, o::PyObject, to_dlpack) where {
+    T, N, A <: AbstractArray{T, N}, M
+}
+    return wrap(A, M, DLManagedTensor(to_dlpack(o)), o)
+end
+
+share(A::StridedArray, from_dlpack::PyObject) = share(A, PyObject, from_dlpack)
```
Comment (rejuvyesh) [inline src/pycall.jl:62]:
```suggestion
share(A::StridedArray, from_dlpack::Union{PyObject,Function}) = share(A, PyObject, from_dlpack)
```

Comment (rejuvyesh) [reply src/pycall.jl:62]: With this fix, https://github.com/rejuvyesh/PyCallChainRules.jl/pull/10 works quite well!

**Comment by rejuvyesh** [comment]
Let me know if you need any help with the tests, but this works quite well! We should ask on `#gpu` to see if we can get access to CUDA buildkite runners for testing.
And again, thank you for all this! The hard part for PyCallChainRules was all here in DLPack :)

File: src/pycall.jl (hunk 2 above)
Comment (pabloferz) [reply src/pycall.jl:62]: If your'e using `PyCall` and a `from_dlpack::Function` you're supposed to be doing `share(A, PyObject, from_dlpack)`. The issue with this change is that if I want to support both PyCall and PythonCall, and someone loads them at the same time, there's going to be ambiguities.

**Comment by pabloferz** [comment]
Thank you @rejuvyesh for testing this over https://github.com/rejuvyesh/PyCallChainRules.jl/pull/10! That guided the design of the interface quite well and helped to make the PR more robust.

**Pull request merged**